reVX/handlers/sc_points.py

```diff
@@ -0,0 +1,618 @@
+        list_in : str | list
+            List or jsonified list from supply curve table
+        dtype : np.dtype
+            dtype for output array
+
+        Returns
+        -------
+        out : ndarary
+            vector of values
+        """
+        if isinstance(list_in, str):
+            list_in = json.loads(list_in)
+        elif not isinstance(list_in, list):
+            msg = ('Must supply either a jsonified list or a list, not {}'
+                   .format(type(list_in)))
+            logger.error(msg)
+            raise ValueError(msg)
+
+        out = np.array(list_in, dtype=dtype)
+
+        return out
+
+    def _drop(self, build_capacity, drop=None):
+        """
+        Drop capacity from Supply Curve point
+
+        Parameters
+        ----------
+        drop : int | None, optional
+            Number of gids to drop, if None drop all, by default None
+
+        Returns
+        -------
+        sc_point : pd.Series
+            Resource gids being allocated
+        capacity : float
+            Capacity being allocated
+        availability : bool
+            Whether Supply Curve point still has available capacity
+        """
+        drop_slice = slice(0, drop, None)
+        capacity = np.sum(self.resource_capacity[drop_slice])
+        if capacity < build_capacity:
+            build_capacity = capacity
+
+        sc_point = {'sc_gid': self.sc_gid,
+                    'res_gids': self.resource_gids[drop_slice].tolist(),
+                    'gid_counts': self.gid_counts[drop_slice].tolist(),
+                    'cf_means': self.cf_means[drop_slice].tolist(),
+                    'build_capacity': build_capacity}
+        sc_point = pd.Series(sc_point)
+
+        if drop is None:
+            self._res_gids = None
+            self._gid_counts = None
+            self._res_capacity = None
+            self._cf_means = None
+            self._capacity = 0.0
+        else:
+            self._res_gids = np.delete(self._res_gids, drop_slice)
+            self._gid_counts = np.delete(self._gid_counts, drop_slice)
+            self._res_capacity = np.delete(self._res_capacity, drop_slice)
+            self._cf_means = np.delete(self._cf_means, drop_slice)
+            self._capacity -= capacity
+
+        if np.isclose(self._capacity, 0):
+            self._capacity = 0.0
+
+        availability = self._capacity > 0
+
+        return sc_point, capacity, availability
+
+    def get_capacity(self, capacity):
+        """
+        Extract capacity from Supply Curve point
+
+        Parameters
+        ----------
+        capacity : float
+            Capacity to extract from Supply Curve point
+
+        Returns
+        -------
+        out : tuple
+            (sc_point, capacity, availability)
+        """
+        build_capacity = capacity.copy()
+        if self.capacity > 0:
+            if capacity < self.capacity:
+                drop = 0
+                for cap in self.resource_capacity:
```

Comment:
I'm surprised that this loop works in conjunction with np.delete() in _drop(). Doesn't np.delete() remove an item from the array? Wouldn't continually removing items from the array and incrementing the drop index result in a broken iteration? see my comment on setting available capacity to zero instead of np.delete().

Reply:
Remove np.delete